Add Feature Flags documentation (#184)
neoforged/Documentation: docs/advanced/featureflags.md

# Feature Flags

Feature Flags are a system which allow developers to gate a set of features behind some set of required flags, that being registered elements, gameplay mechanics, data pack entries or some other unique system to your mod.

> A common use case would be gating experimental features/elements behind a experimental flag, allowing users to easily switch them on and play around with them before they are finalized.

Author: what do you mean here exactly?
do you mean all the pale oak blocks are gated cause they depend on the pale oak wood type?
if so thats just simply not true, `WoodType` and its paired `BlockSetType` cannot be gated behind flags, its each individual `Block` thats defining the flag thats gating them, which would fall under `registered elements,`.

> A common use case would be gating experimental features/elements behind a experimental flag, allowing users to easily switch them on and play around with them before they are finalized.

Reviewer: No, I just wanted something like "You can also gate your own blocks behind vanilla feature flags. For example if you were to add to the set of Pale Oak wood blocks, you'd only want those to show up if Pale Oak is enabled." (but more general because using Pale Oak as an example is gonna be outdated by the next version)

DeferredHolder<EntityType<?>, EntityType<ExperimentalEntity>> EXPERIMENTAL_ENTITY = ENTITY_TYPES.register("experimental", registryName -> EntityType.Builder.of(ExperimentalEntity::new, MobCategory.AMBIENT)
    .requiredFeatures(EXPERIMENTAL) // mark as requiring the 'EXPERIMENTAL' flag
    .build(ResourceKey.create(Registries.ENTITY_TYPE, registryName))
);

// MenuType
DeferredRegister<MenuType<?>> MENU_TYPES = DeferredRegister.create(Registries.MENU, "examplemod");
DeferredHolder<MenuType<?>, MenuType<ExperimentalMenu>> EXPERIMENTAL_MENU = MENU_TYPES.register("experimental", () -> new MenuType<>(
    // Using Vanilla MenuSupplier
    // This is used when your menu is not encoding complex data during `player.openMenu`
    // (windowId, inventory) -> new ExperimentalMenu(windowId, inventory),

    // Using NeoForge IContainerFactory
    // This is used when you wish to read complex data encoded during `player.openMenu`
    // Casting here is important as `MenuType` specificly wants `MenuSupplier`
    (IContainerFactory<ExperimentalMenu>) (windowId, inventory, buffer) -> new ExperimentalMenu(windowId, inventory, buffer),
    
    // mark as requiring the 'EXPERIMENTAL' flag

>     FeatureFlagSet.of(EXPERIMENTAL)

Author: Im confused what you mean here? do you mean remove the `FeatureFlagSet.of`?
To match everything else which is just `.requiredFeatures(EXPERIMENTAL)`
This cant be done as `MenuType`s ctor specifically wants a `FeatureFlagSet` where as everything else wants `FeatureFlag...` which then is wrapped into a `FeatureFlagSet.of` call

Or do you mean put the comment on the same line as the code?
I put the comment above code here to match the lines above, but i can move the comment if thats desired

>     FeatureFlagSet.of(EXPERIMENTAL)

Reviewer: Yeah, this is mainly about the comment. Above you always use `.requiredFeatures(EXPERIMENTAL) // comment`, so this should similarly be `FeatureFlagSet.of(EXPERIMENTAL) // comment`

To validate if any given `FeatureElement` is enabled, simply invoke `FeatureElement#isEnabled`.

:::note
`ItemStack` has a speciality `isItemEnabled(FeatureFlagSet)` method, This is so that empty stacks are treated as enabled even if the required features for backing `Item` does not match the enabled features.

It is recommended to make use of this speciality over `Item#isEnabled` where possible.
:::

```java
requiredFeatures.isSubsetOf(enabledFeatures);
featureElement.isEnabled(enabledFeatures);
itemStack.isItemEnabled(enabledFeatures);
```

### Feature Packs

Feature Packs are a type of pack which not only loads resources and/or data but also has the ability to toggle on a given set of Feature Flags.

These flags are defined in the `pack.mcmeta` JSON file at the root of this pack, which follows the below format.
```json5
{
    "features": [
        "enabled": [
            // Identifier of a Feature Flag to be enabled
            // Must be a valid registered flag
            "examplemod:experimental"
        ]
    ]
}
```

There are a couple methods in which you can provide a feature pack to your users.

> Installing these packs differs depending on if its being installed into a single player world or a server.

Reviewer: I don't really understand these sentences in context. Do you mean that there's different ways of installing (built-in vs. external, and client vs. server)? If so, state this here directly, for example like so:

"There are a couple of ways for users to obtain a feature pack, namely installing them from an external source as a datapack, or downloading a mod that has a built-in feature pack. Both of these then need to be installed differently depending on the physical side." Also link to the Physical Sides section in the Sides article.

**External**

External packs are provided to your users in datapack form.

> **Installing into a single player world**

Reviewer: ```suggestion
#### Installation in Singleplayer
```
Make this a H4 since it is a subsection of External, which would be a H3. Also add a new line after the heading.

1. Create a new world.
2. Navigate to the datapack selection screen.
3. Drag and drop the datapack zip file onto game window.
4. Move the newly available datapack over to the `Selected` packs list.
5. Confirm changes by clicking `Done`.

The game will now warn you about any newly selected experimental features, potential bugs, issues and crashes.
You can confirm these changes by clicking `Proceed` or `Details` to see an extensive list of all selected packs and which features they would enable.

:::note
External Feature Packs do not show up in the Experiments screen.

To disable these packs after enabling them, navigate back into the datapacks screen and move the external packs back into `Available` from `Selected`.

The Experiments screen only allows toggling on or off Built-In Feature Packs.
:::

> **Installing into a server**

Reviewer: ```suggestion
#### Installation in Multiplayer
```

Enabling Feature Packs can only be done during initial world creation, and cannot be disabled once enabled.

1. Create the directory `./world/datapacks`
2. Upload the datapack zip file into the newly created directory
3. Open your servers `server.properties` file

> 4. Add the datapack zip file name (excluding `.zip`) to `initial-enabled-packs` (separating each pack by a `,`)

Reviewer: Maybe add an example here.

> The Feature Pack is now installed and enabled

Reviewer: This line seems unnecessary to me.

**Built-In**

> Built-In packs are bundled with your mod and are made available to the game using the `AddPackFindersEvent` event.

Reviewer: ```suggestion
Built-in packs are bundled with your mod and are made available to the game using the `AddPackFindersEvent` event.
```

```java
@SubscribeEvent
public static void addFeaturePacks(final AddPackFindersEvent event) {
    event.addPackFinders(

>             // 'packLocation': Path relative to your mods 'resources' pointing towards this pack

Reviewer: The parameter name should not be added to the comment, an explanation of what is expected here should be enough.

            ResourceLocation.fromNamespaceAndPath("examplemod", "data/examplemod/datapacks/experimental"),
            
            // 'packType': What kind of resources are contained within this pack
            // 'CLIENT_RESOURCES' for packs with client assets (resource packs)
            // 'SERVER_DATA' for packs with server data (data packs)
            PackType.SERVER_DATA,
            
            // 'packNameDisplay': Display name shown in the Experiments screen
            Component.literal("ExampleMod: Experiments"),
            

>             // 'packSource': In order for this pack to load and enable feature flags this *MUST* be 'FEATURE'
>             // any of 'PackSource' type is invalid here

Reviewer: ```suggestion
            // In order for this pack to load and enable feature flags, this MUST be 'FEATURE',
            // any other PackSource type is invalid here
```

            PackSource.FEATURE,
            

>             // 'alwaysActive': If 'true' this pack is always active and cannot be disabled, should always be 'false' for feature packs

Reviewer: ```suggestion
            // If this is true, the pack is always active and cannot be disabled, should always be false for feature packs
```
